## Problemas Estatística- 22/09

In [1]:
print("hello world")

hello world


### Problema 1- Melhor Desempenho

In [2]:
import statistics 

In [5]:
def calcular_modas(valores):
    modas = []
    maior_frequencia = 0

    for valor in valores:
        frequencia = valores.count(valor)

        if frequencia > maior_frequencia:
            maior_frequencia = frequencia
            modas = [valor]

        elif frequencia == maior_frequencia and valor not in modas:
            modas.append(valor)

    if maior_frequencia == 1:
        return "Amodal", []
    elif len(modas) == 1:
        return "Unimodal", modas
    elif len(modas) == 2:
        return "Bimodal", modas
    else:
        return "Multimodal", modas

In [6]:
def calcular_amplitude(lista):
    amplitude = max(lista) - min(lista)
    return amplitude

In [ ]:
def medidas_analise(lista):
    media = statistics.mean(lista)
    mediana = statistics.median(lista)
    moda = calcular_modas(lista)
    amplitude = calcular_amplitude(lista)

    return media, mediana, moda, amplitude


In [12]:
equipe_A = [42,38,45,41,39,44,38]
media_A, mediana_A, moda_A, amplitude_A = medidas_analise(equipe_A)

equipe_B = [30,52,35,48,29,55,38]
media_B, mediana_B, moda_B, amplitude_B = medidas_analise(equipe_B)

print("Resultados Encontrados:\n")
print(f"Equipe A: {equipe_A}\n Média: {media_A}\n Mediana: {mediana_A}\n Moda: {moda_A}\n Amplitude: {amplitude_A}")
print("\n-----------------------------\n")
print(f"Equipe B: {equipe_B}\n Média: {media_B}\n Mediana: {mediana_B}\n Moda: {moda_B}\n Amplitude: {amplitude_B}")

Resultados Encontrados:

Equipe A: [42, 38, 45, 41, 39, 44, 38]
 Média: 41
 Mediana: 41
 Moda: ('Unimodal', [38])
 Amplitude: 7

-----------------------------

Equipe B: [30, 52, 35, 48, 29, 55, 38]
 Média: 41
 Mediana: 38
 Moda: ('Amodal', [])
 Amplitude: 26


#### CONCLUSÃO:

A Equipe A possui melhor desempenho, seus valores de medidas centrais estão mais semelhantes entre si, além de que sua amplitude é menor não tendo uma grande variação, então ela possui uma centralidade melhor e com menos desvio. Já a Equipe B trás alguns dados discrepantes, sua amplitude é elevada sendo de 26 e o conjunto é Amodal, sendo assim o conjunto B ele não está centralizado com valores dispersos no conjunto.

### Problema 2- Máquina Regular 

In [18]:
def calcular_variancia(lista):
    qnt_lista = len(lista)
    media = statistics.mean(lista)
    somatoria = 0

    for elemento in lista:
        diferenca_quadrado = pow((elemento - media),2)
        somatoria += diferenca_quadrado
        
    variancia = somatoria / qnt_lista

    return variancia

In [19]:
import math 
def calcular_desvio_padrao(lista):
    variancia = calcular_variancia(lista)
    desvio_padrao = math.sqrt(variancia)

    return desvio_padrao

In [24]:
def calcular_cof_variacao(lista):
    media = statistics.mean(lista)
    dp = calcular_desvio_padrao(lista)

    coeficiente_variacao = (dp / media) * 100

    return coeficiente_variacao

In [25]:
def medidas_desvio(lista):
    media = statistics.mean(lista)
    variancia = calcular_variancia(lista)
    desvio = calcular_desvio_padrao(lista)
    amplitude = calcular_amplitude(lista)
    cof_variacao = calcular_cof_variacao(lista)

    return media, variancia, desvio, amplitude, cof_variacao

In [27]:
maquina_A = [98,101,100,99,102,100,101]
media_A, variancia_A, desvio_A, amplitude_A, cof_A = medidas_desvio(maquina_A)

maquina_B = [82,115,91,108,120,86,99]
media_B, variancia_B, desvio_B, amplitude_B, cof_B = medidas_desvio(maquina_B)

print("Resultados Encontrados:\n")
print(f"Maquina A: {maquina_A}\n Média: {media_A:.2f}\n Variância: {variancia_A:.2f}\n Desvio: {desvio_A:.2f}\n Amplitude: {amplitude_A}\n Coeficiente de Variação: {cof_A:.2f}%")
print("\n-----------------------------\n")
print(f"Maquina B: {maquina_B}\n Média: {media_B:.2f}\n Variância: {variancia_B:.2f}\n Desvio: {desvio_B:.2f}\n Amplitude: {amplitude_B}\n Coeficiente de Variação: {cof_B:.2f}%")

Resultados Encontrados:

Maquina A: [98, 101, 100, 99, 102, 100, 101]
 Média: 100.14
 Variância: 1.55
 Desvio: 1.25
 Amplitude: 4
 Coeficiente de Variação: 1.24%

-----------------------------

Maquina B: [82, 115, 91, 108, 120, 86, 99]
 Média: 100.14
 Variância: 184.41
 Desvio: 13.58
 Amplitude: 38
 Coeficiente de Variação: 13.56%


#### CONCLUSÃO

A Maquina A apresenta uma produção mais regular, a média de ambas as máquinas ficam no valor de 100,14, no entanto a Máquina A possui a variância populacional e  desvio padrão menor, sendo, 1,55 e 1,256 respectivamente, assim mostrando que não há uma grande varição dos valores. Diferentemente da máquina B que possui valores distintos e seu desvio está em 13,58, apresentando grande variabilidade. E a amplitude das máquinas apenas comprova esse fato, enquanto da máquina A se estabelece em 4, tendo valores semelhantes e consistentes, o da máquina B está em 38, apresentando um valor alto de amplitude. O coeficiente de variação do A é de 1,24% já da B é de 13,56% sendo ela a irregular.

### Problema 3- Tempos de Entrega 

In [28]:
def calcular_quartis_otimizado(lista):

    n = len(lista)
    lista_ordenada = sorted(lista)

    posicoes = [
        (n + 1) / 4,
        (n + 1) / 2,
        (3 * (n + 1)) / 4
    ]

    quartis = []

    for posicao in posicoes:
        posicao_inteira = int(posicao)

        if posicao == posicao_inteira:
            valor_quartil = lista_ordenada[posicao_inteira - 1]

        else:
            parte_decimal = posicao - posicao_inteira

            diferenca = lista_ordenada[posicao_inteira] - lista_ordenada[posicao_inteira - 1]
            proporcao = diferenca * parte_decimal
            valor_quartil = lista_ordenada[posicao_inteira - 1] + proporcao

        quartis.append(valor_quartil)

    return quartis

In [31]:
def calcular_intervalo_quartilico(lista):
    q1, q2, q3 = calcular_quartis_otimizado(lista)
    intervalo = q3 - q1

    return intervalo

In [32]:
def medidas_quartis(lista):
    media = statistics.mean(lista)
    mediana = statistics.median(lista)
    variancia = calcular_variancia(lista)
    desvio = calcular_desvio_padrao(lista)
    amplitude = calcular_amplitude(lista)
    quartis = calcular_quartis_otimizado(lista)
    intervalo = calcular_intervalo_quartilico(lista)

    return media, mediana, variancia, desvio, amplitude, quartis, intervalo


In [37]:
tempos = [48,32,55,41,37,62,29,45,51,35,70]
media, mediana, variancia, desvio, amplitude, quartis, intervalo = medidas_quartis(tempos)

print("Resultados Encontrados:\n")
print(f"Tempos: {tempos}\n Média: {media:.2f}\n Mediana: {mediana}\n"
        f" Variância: {variancia:.2f}\n Desvio: {desvio:.2f}\n Amplitude: {amplitude}\n"
        f" Quartis: {quartis}\n Intervalo Quartilico: {intervalo}")

Resultados Encontrados:

Tempos: [48, 32, 55, 41, 37, 62, 29, 45, 51, 35, 70]
 Média: 45.91
 Mediana: 45
 Variância: 150.45
 Desvio: 12.27
 Amplitude: 41
 Quartis: [35, 45, 55]
 Intervalo Quartilico: 20


#### CONCLUSÃO

Em relação a afirmação "A maioria das entregas ocorre muito próxima da média" está errado, os valores estão próximos, sendo mais da metade dos valores estão próximos, no entanto afirmar que está muito próximo está incorreto, pois existem valores bem distantes, como o 32 e 70, que seriam dois outliers, levqando em consideração que a média é estabelecida em 45.91

### Problema 4- Análise de vendas por vendedor 

In [1]:
import statistics

In [4]:
def registrar_vendas(vendedor, *vendas):
    print(f"O vendedor {vendedor}, teve {len(vendas)} vendas")
    valor_total= sum(vendas)
    media_vendas = statistics.mean(vendas)
    maximo_valor = max(vendas)

    print(f"Valor total vendido: {valor_total}\nMedia do valor de vendas: {media_vendas}\nValor máximo vendido: {maximo_valor}")

    if valor_total >= 5000:
        print("Meta atingida!")
    else:
        print("Meta NÃO atingida!")

In [5]:
registrar_vendas(
    "José",
    2000,
    100,
    112,
    500,
    200,
    990,
    1100,
    1500
)

O vendedor José, teve 8 vendas
Valor total vendido: 6502
Media do valor de vendas: 812.75
Valor máximo vendido: 2000
Meta atingida!


In [7]:
registrar_vendas(
    "Pedro",
    500,
    1700,
    210,
    140,
    800
)

O vendedor Pedro, teve 5 vendas
Valor total vendido: 3350
Media do valor de vendas: 670
Valor máximo vendido: 1700
Meta NÃO atingida!


In [9]:
registrar_vendas(
    "Lucas",
    2000,
    2000,
    1600,
    200
)

O vendedor Lucas, teve 4 vendas
Valor total vendido: 5800
Media do valor de vendas: 1450
Valor máximo vendido: 2000
Meta atingida!


### Problema 5- Registro de atendimento 

In [44]:
def registrar_atendimentos(paciente, *procedimentos, **informacoes):

    print(f"Paciente {paciente} tem {len(procedimentos)} procedimentos\n")

    print(50*"-")
    print("Procedimentos:")
    for i, proced in enumerate(procedimentos, start=1):
        print(f"{i}- {proced}")
    print(50*"-")

    print()
    print(50*"-")
    print(f"Informações do paciente:")
    retorno = False
    for chave, valor in informacoes.items():
        print(f"{chave}: {valor}")
        if chave == 'data_retorno':
            retorno= True 
    print(50*"-")
    
    if retorno:
        print(f"\nRetorno agendado para data: {informacoes['data_retorno']}")
    else:
        print("\nRetorno não informado!")

In [45]:
registrar_atendimentos(
    "Victor",
    "Botox",
    "Bariátrica",
    "Silicone",
    idade= 21,
    observacoes= "Diabético"
)

Paciente Victor tem 3 procedimentos

--------------------------------------------------
Procedimentos:
1- Botox
2- Bariátrica
3- Silicone
--------------------------------------------------

--------------------------------------------------
Informações do paciente:
idade: 21
observacoes: Diabético
--------------------------------------------------

Retorno não informado!


In [46]:
registrar_atendimentos(
    "Felipe",
    "Consulta geral",
    "Exame de sangue",
    profissional_responsavel= "Lapa",
    convenio= True,
    data_retorno= "10/10/2030"
)

Paciente Felipe tem 2 procedimentos

--------------------------------------------------
Procedimentos:
1- Consulta geral
2- Exame de sangue
--------------------------------------------------

--------------------------------------------------
Informações do paciente:
profissional_responsavel: Lapa
convenio: True
data_retorno: 10/10/2030
--------------------------------------------------

Retorno agendado para data: 10/10/2030


In [47]:
registrar_atendimentos(
    "Julia",
    "Clareamento dental"
)

Paciente Julia tem 1 procedimentos

--------------------------------------------------
Procedimentos:
1- Clareamento dental
--------------------------------------------------

--------------------------------------------------
Informações do paciente:
--------------------------------------------------

Retorno não informado!
